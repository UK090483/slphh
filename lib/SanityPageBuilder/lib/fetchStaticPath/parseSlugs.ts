import {
  LocationConfig,
  FetchStaticPathsResult2,
} from "@lib/SanityPageBuilder/types";
import { defaultLocales } from "./defaultValues";
import { GetSlugsResult } from "./getSlugs";

export type ParseSlugsProps = {
  getSlugResult: GetSlugsResult;
  locales?: LocationConfig;
  fallback?: boolean | "blocking";
};

const parseSlugs = (props: ParseSlugsProps): FetchStaticPathsResult2 => {
  const {
    getSlugResult,
    locales = defaultLocales,
    fallback = "blocking",
  } = props;

  if (
    !getSlugResult ||
    !getSlugResult.slugs ||
    !Array.isArray(getSlugResult.slugs)
  ) {
    throw new Error("parseSlugs did not get useful data on  getSlugResult");
  }
  const hasI18n = Object.keys(locales).length > 1;
  const res = {
    paths:
      getSlugResult.slugs.reduce((acc, page) => {
        if (!page.slug) return [...acc];

        let pageParams = Object.entries(locales).map(([locale, item]) => {
          let slug = item.isDefault
            ? page.slug
            : (page[`slug-${locale}`] as string) || page.slug;

          return {
            params: {
              slug: slug.split("/").map((i) => i.toLowerCase()),
            },
            ...(hasI18n && { locale }),
          };
        });
        return [...acc, ...pageParams];
      }, [] as any[]) || [],
    fallback,
  };

  return res;
};

export default parseSlugs;
