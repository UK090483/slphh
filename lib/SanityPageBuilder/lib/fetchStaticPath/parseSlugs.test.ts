import parseSlugs from "./parseSlugs";

const getSlugResult = {
  slugs: ["slug1", "slug/split"].map((slug) => ({
    slug,
    "slug-en": `${slug}-en`,
  })),
};

const locales = {
  de: { title: "Deutsch", isDefault: true, flag: "ðŸ‡©ðŸ‡ª" },
  en: { title: "Englisch", flag: "ðŸ‡ºðŸ‡¸" },
  da: { title: "DÃ¤nisch", flag: "ðŸ‡©ðŸ‡°" },
};

describe("parseSlugs", () => {
  it("should trow error if fetch result is not usable ", () => {
    let didThrow = false;
    try {
      //@ts-ignore
      parseSlugs({ getSlugResult: null });
    } catch (error) {
      didThrow = true;
    }
    expect(didThrow).toBe(true);
  });

  it("should return expected result no locales", () => {
    expect(parseSlugs({ getSlugResult })).toMatchSnapshot();
  });
  it("should return expected result with locales", () => {
    expect(parseSlugs({ getSlugResult, locales })).toMatchSnapshot();
  });
});
