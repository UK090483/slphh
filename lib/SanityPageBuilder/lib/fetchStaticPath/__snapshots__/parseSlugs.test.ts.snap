// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseSlugs should return expected result no locales 1`] = `
Object {
  "fallback": "blocking",
  "paths": Array [
    Object {
      "params": Object {
        "slug": Array [
          "slug1",
        ],
      },
    },
    Object {
      "params": Object {
        "slug": Array [
          "slug",
          "split",
        ],
      },
    },
  ],
}
`;

exports[`parseSlugs should return expected result with locales 1`] = `
Object {
  "fallback": "blocking",
  "paths": Array [
    Object {
      "locale": "de",
      "params": Object {
        "slug": Array [
          "slug1",
        ],
      },
    },
    Object {
      "locale": "en",
      "params": Object {
        "slug": Array [
          "slug1-en",
        ],
      },
    },
    Object {
      "locale": "da",
      "params": Object {
        "slug": Array [
          "slug1",
        ],
      },
    },
    Object {
      "locale": "de",
      "params": Object {
        "slug": Array [
          "slug",
          "split",
        ],
      },
    },
    Object {
      "locale": "en",
      "params": Object {
        "slug": Array [
          "slug",
          "split-en",
        ],
      },
    },
    Object {
      "locale": "da",
      "params": Object {
        "slug": Array [
          "slug",
          "split",
        ],
      },
    },
  ],
}
`;
